Draft a deployment notifier that posts on PR pre-prod status (#32459)
sourcegraph/sourcegraph: enterprise/dev/deployment-notifier/README.md

# Deployment Notifier

Deployment Notifier is a tool that analyzes changes in order to guess which pull requests have been deployed in a given environment.
It is meant to be included in deploymement pipeline to automatically notify teammates that their changes have been deployed.

## Usage

In a deployment repository (such as `sourcegraph/deploy-sourcegraph-*`) you can run the following command to post GitHub comments and Slack notifications.

```sh
deployment-notifier -environment $MY_ENV -slack.token=$SLACK_TOKEN -slack.webhook=$SLACK_WEBHOOK
```

### Flags

- `-github.token` (defaults to `$GITHUB_TOKEN`)
- `-environment` either `preprod` or `production`
- `-dry` (optional) do not post on Slack or GitHub, just print out what would be posted.
- `-slack.token` Slack Token used to find the matching Slack handle for pull request authors.
- `-slack.webhook` Slack webhook URL to post the notifications on.

## How it works

Deployment notifier works as following:

> 1. Inspect the diff of current commit (supposed to be a merge commit)

Reviewer: ```suggestion
1. Inspect the diff the deployment manifests' current commit (supposed to be a merge commit, e.g. run on a release or deploy branch)
```

> 2. Compute the deployed applications and take note of the old commit and new commit for each of them.

Reviewer: ```suggestion
2. Compute the deployed applications and take note of the old and new Sourcegraph commit for each of them.
```

> 3. For each set of old and new commits, find all PRs that happened after the old commit, up to the new commit.

Reviewer: ```suggestion
3. For each set of old and new Sourcegraph commits, find all PRs that happened after the old commit, up to the new commit.
```

4. Post a comment in each of those PRs with the list of applications computed in 2.
5. Post a Slack message pinging all pull request authors.

## Contributing

### Local usage

To avoid spamming comments, Deployment Notifier comes with a few flags to help:

> - `-dry` prints a summary of what would be posted on the standard output.

Reviewer: 👨‍🍳
